haystack_experimental/utils/jinja_chat_extension.py

```diff
@@ -0,0 +1,203 @@
+        """
+        Parse the message tag and its attributes in the Jinja2 template.
+
+        This method handles the parsing of role (mandatory), name (optional), meta (optional) and message body content.
+
+        :param parser: The Jinja2 parser instance
+        :return: A CallBlock node containing the parsed message configuration
+        :raises TemplateSyntaxError: If an invalid role is provided
+        """
+        lineno = next(parser.stream).lineno
+
+        # Parse role attribute (mandatory)
+        parser.stream.expect("name:role")
+        parser.stream.expect("assign")
+        role_expr = parser.parse_expression()
+
+        if isinstance(role_expr, nodes.Const):
+            role = role_expr.value
+            if role not in self.SUPPORTED_ROLES:
+                raise TemplateSyntaxError(f"Role must be one of: {', '.join(self.SUPPORTED_ROLES)}", lineno)
+
+        # Parse optional name attribute
+        name_expr = None
+        if parser.stream.current.test("name:name"):
+            parser.stream.skip()
+            parser.stream.expect("assign")
+            name_expr = parser.parse_expression()
+
+        # Parse optional meta attribute
+        meta_expr = None
+        if parser.stream.current.test("name:meta"):
+            parser.stream.skip()
+            parser.stream.expect("assign")
+            meta_expr = parser.parse_expression()
+
+        # Parse message body
+        body = parser.parse_statements(("name:endmessage",), drop_needle=True)
+
+        # Build message node with all parameters
+        return nodes.CallBlock(
+            self.call_method(
+                name="_build_chat_message_json",
+                args=[
+                    role_expr,
+                    name_expr or nodes.Const(None),
+                    meta_expr or nodes.Dict([]),
+                ],
+            ),
+            [],
+            [],
+            body,
+        ).set_lineno(lineno)
+
+    def _build_chat_message_json(self, role: str, name: Optional[str], meta: dict, caller: Callable[[], str]) -> str:
+        """
+        Build a ChatMessage object from template content and serialize it to a JSON string.
+
+        This method is called by Jinja2 when processing a {% message %} tag.
+        It takes the rendered content from the template, converts XML blocks into ChatMessageContentT objects,
+        creates a ChatMessage object and serializes it to a JSON string.
+
+        :param role: The role of the message
+        :param name: Optional name for the message sender
+        :param meta: Optional metadata dictionary
+        :param caller: Callable that returns the rendered content
+        :return: A JSON string representation of the ChatMessage object
+        :raises ValueError: If the message content is empty
+        """
+
+        content = caller()
+        parts = self._parse_content_parts(content)
+        if not parts:
+            raise ValueError("Message content is empty")
```

Comment:
Would it be possible to provide some additional contextual information here? Like print out a snippet of the incorrect Jinja2 template and/or a line number? I'm just imagining for long templates or when you have multiple ChatPromptBuilders in a pipeline it could be hard to go find the wrongly formatted template based on this error message alone.

Reply:
I've tried to improve the checks for empty content but now most errors are triggered because the content is empty or only consists of whitespace chars. I cannot trigger an error for non-empty content producing 0 content parts.
https://github.com/deepset-ai/haystack-experimental/pull/297/commits/ebf084d482cc0f393ad190c66e1c8c678f5d4903